### PR TITLE
Remove warnings

### PR DESCRIPTION
**Pull request by jokva**
This PR accomplishes two things:

1. It increases the warning level significantly. The previous warning level wasn't properly set because `CMAKE_CXX_FLAGS` typically was defined. The warnings have been made default by unconditionally setting them (when not microsoft), but only in debug mode. Debug mode is now the default build configuration.
2. I've gone through all warnings generated by this new warning set and fixed them.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -77,20 +77,19 @@ endif (USE_RUNPATH)
 
 if (MSVC)
   add_definitions( "/W3 /D_CRT_SECURE_NO_WARNINGS /wd4996 /wd4244 /wd4267")
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /MP")
+  set(CMAKE_CXX_FLAGS "/MP ${CMAKE_CXX_FLAGS}")
 else()
-  if (NOT DEFINED CMAKE_CXX_FLAGS)
```
Comment (blattms) [inline CMakeLists.txt]: We should give people the chance to overwrite these flags.

Maybe introducing a CACHE variable OPM_CMAKE_CXX_FLAGS_DEBUG and and adding them to CMAKE_CXX_FLAGS_DEBUG is an option
```
set(OPM_CMAKE_CXX_FLAGS_DEBUG  "-pipe -std=c++0x -pedantic -Wall -Wextra -Wformat-nonliteral -Wcast-align -Wpointer-arith -Wmissing-declarations -Wcast-qual -Wshadow -Wwrite-strings -Wchar-subscripts -Wredundant-decls"
   CACHE STRING "Default flags for OPM")
set(CMAK_CXXFLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} ${OPM_CMAKE_CXX_FLAGS_DEBUG}")
```
IMHO the same should be done for the other flags (RELEASE, DEBUG), too.

By the way, what happens if we rerun CMake? Do they get duplicated? I guess that might have been  a reason for the former  if statement.

Comment (jokva) [inline CMakeLists.txt]: They get to override them now (by passing the respective -Wno or -std= flags).

I can try running cmake twice, give me a second.

Comment (jokva) [inline CMakeLists.txt]: Running cmake twice doesn't seem to duplicate the list.

**Comment by joakim-hove** [comment]
Merge?

**Pull request merged**